### ISSUE TITLE
Mounting on a subdomain is broken

### ISSUE THREAD

**Issue by checkbutton** [hanami/hanami#951]
Specifying `host:` when mounting applications in Hanami seems to be broken.
I have the following setup:
```
Hanami.configure do
  mount TestOne::Application, at: '/', host: 'test-one'
  mount TestTwo::Application, at: '/', host: 'test-two'
# ...
end
```
In both applications, I define the following route:
```
get '/hello', to: ->(env) { [200, {}, ['Hello from Test One/Two!']] }
```
When I run my application with `hanami server --host=lvh.me`, as described in http://hanamirb.org/guides/1.2/routing/overview/#mounting-on-a-subdomain and try to call `test-two.lvh.me/hello`, I end up in TestOne.

Moreover, if I stick to the documentation and leave out the `at:` in mount, I end up getting
```
key not found: :at
/Users/user/.rvm/gems/ruby-2.5.1/gems/hanami-1.2.0/lib/hanami/configuration.rb:35:in `fetch'
```

Setting `at: ` to an empty String, it does not build any route (only getting 404s).

**Comment by cllns** [comment]
Sorry you're having trouble with this! I think the documentation might be wrong. Can you try `  mount TestOne::Application, at: '/', host: 'test-one.lvh.me'`? I know this isn't idea, since the root host will chance per environment, but I seem to remember it needed the full host name, not just the subdomain.

**Comment by checkbutton** [comment]
@cllns thanks for your answer. I've just tried that out, but without any effect. I've also tried to set the `host` in the respective `TestOne::Application` / `TestTwo::Application` `configure` block, but this as well had no effect.

**Comment by checkbutton** [comment]
Additional info: I tried it also with different web servers, i.e. WEBrick and Puma

**Comment by greggilbert** [comment]
I'm also having this issue.

[The docs](http://hanamirb.org/guides/1.2/routing/overview/) note that you can do `mount Blog.new, host: 'blog'`, but using the `.new` notation throws this:
```

**Comment by greggilbert** [comment]
I'm also having this issue.

Per [the docs](http://hanamirb.org/guides/1.2/routing/overview/), one notation for this is to use `.new`, as in `mount Blog.new, host: 'blog'`. When you do that, though, it throws this:

```
NoMethodError: undefined method `controller_pattern' for nil:NilClass

/usr/local/bundle/gems/hanami-1.2.0/lib/hanami/rendering_policy.rb:32:in `initialize'
/usr/local/bundle/gems/hanami-1.2.0/lib/hanami/application.rb:152:in `new'
/usr/local/bundle/gems/hanami-1.2.0/lib/hanami/application.rb:152:in `initialize'
/usr/src/app/config/environment.rb:10:in `new'
```

I've also noticed that order is significant here, with the first mount line winning out:
```
Hanami.configure do
  mount Web::Application, at: '/'    # <-- the web routes win, and Admin is inaccessible
  mount Admin::Application, at: '/'
```

I'm going to keep digging, but this seems like a big issue, especially since hanami/router appears to be able to handle it.

**Comment by AlfonsoUceda** [comment]
Hi Folks, sorry for the problem but yes, it is a bug.

I've reproduced the bug and I know how to fix it, so I'll create a PR, after that can you test it in your apps? thanks!

**Comment by AlfonsoUceda** [comment]
@checkbutton @greggilbert I've created a quick PR that it needs to be improved. https://github.com/hanami/hanami/pull/976

Can you test with your app? Thanks

**Comment by glaszig** [comment]
i can confirm that the patch from #976 (without the merging of mount host + env host, though) is working with the following configuration. i have this running in "production", whatever that really means.

```nginx
# nginx.conf
server {
  server_name beta.example.com www.example.com;
  location / {
    proxy_set_header Host              $http_host;
    proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host  $http_host:$server_port;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_pass http://hanami_app;
}
```

```rb
# hanami_app/config/environment.rb
Hanami.configure do
  mount Beta::Application, at: '/', host: "beta.example.com"
  mount Admin::Application, at: '/admin'
  mount Web::Application, at: '/'
end
```

- there's 2 apps mounted at `/` but one of them is supposed to only respond to requests with the `Host: beta.example.com` request header
- the other two apps, `Web` and `Admin`, respond to all other requests -- in this case `Host: www.example.com`

**Comment by glaszig** [comment]
i would also like to point out that the `--host` argument to the `hanami server` cli command _does not_ have very much to do with this. this argument simply allows you to define on which _interface_ on your machine where you are running the hanami server you'd like to bind the socket.

```
$ hanami server --help
...
  --host=VALUE                    	# The host address to bind to
```

it says right there "address to **bind**" to. this works with `--host foo.lvh.me` only because lvh resolves everything to `127.0.0.1`. if you try and address for which your the machine does not have an interface, e.g., `3.4.5.6`, you'll get an exception.

on the other hand, a value of `0.0.0.0` binds the socket to all available interfaces and makes your that socket be reachable from every address space you have interfaces for.
